### PR TITLE
Implementation of LSTM Model

### PR DESCRIPTION
**Pull request by frankfka**
# Pull Request

## Description

- Used model class to create LSTM model
- Fitting takes around 5 min depending on the computer - again this depends on the epochs used and the number of sequences that we train with
- Current implementation works, but uses a lot of arbitrary constants, we should look at building out the tokenization & fitting functionality to get better predictions.

## Testing
- We should look at making predictable tests for this model class. I tested with a small test script that isn't included in the PR. 

## Ticket(s)

Affects #10

### CODE REVIEW

File: mimic/model/lstm_model.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+from mimic.model.model import Model
+
+from keras.preprocessing.sequence import pad_sequences
```
Comment (rwblickhan) [inline mimic/model/lstm_model.py]: Keras and TensorFlow will have to be added to the Pipfile once #14 is merged.

File: mimic/model/lstm_model.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import numpy as np
+import string, os 
+
+class LSTMModel(Model):
+
+    def __init__(self):
+        self.tokenizer = Tokenizer()
+
+    ''' Create sequences to train the model - each is a sequence of tokens that represent characters '''
```
Comment (rwblickhan) [inline mimic/model/lstm_model.py]: You'll also want to revisit this once #14 is merged - `pydocstyle` is quite picky about how you write docstrings.

File: mimic/model/lstm_model.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def learn(self, text):
+
+        # TODO These are currently arbitrary - we can look at varying these depending on the size of the input text
+        SEQ_LEN = 100
+        BATCH_SIZE = 200
+
+        # Clean text of weird characters
+        text = "".join(v for v in text if v not in string.punctuation).lower()
+        clean_txt = text.encode("utf8").decode("ascii",'ignore')
```
Comment (rwblickhan) [inline mimic/model/lstm_model.py]: Do you think this would also be valuable for other models (e.g. Markov chain)? If so, we might want to refactor this into a cleaning/preprocessing module (that would presumably be called by the generator when loading texting).

Comment (rwblickhan) [reply mimic/model/lstm_model.py]: (That being said, I wouldn't bother changing it for this PR - we can make a new ticket to change it.)

File: mimic/model/lstm_model.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        model.add(Dense(total_words, activation='softmax'))
+        model.compile(loss='categorical_crossentropy', optimizer='adam')
+        model.fit(predictors, label, epochs=100, verbose=5)
+
+        self.max_sequence_len = max_sequence_len
+        self.model = model 
+
+    def predict(self):
+        # TODO this uses arbitrary constants for now
+        seed_text = "where art thou"
```
Comment (rwblickhan) [inline mimic/model/lstm_model.py]: Is seed text necessary for this to work? What will we use as seed text in "normal operation"?

File: mimic/model/lstm_model.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import numpy as np
+import string, os 
+
+class LSTMModel(Model):
+
+    def __init__(self):
+        self.tokenizer = Tokenizer()
+
+    ''' Create sequences to train the model - each is a sequence of tokens that represent characters '''
+    def learn(self, text):
```
Comment (rwblickhan) [inline mimic/model/lstm_model.py]: What format should `text` be in? It might be worthwhile for @Annnnnieee and @jackattackyang to take a look at this as well so we're all on the same page.

File: mimic/model/lstm_model.py (hunk 3 above)
Comment (frankfka) [reply mimic/model/lstm_model.py]: Yeah, most likely. I was thinking to do so but we haven't discussed how we were going to handle util methods yet

File: mimic/model/lstm_model.py (hunk 4 above)
Comment (frankfka) [reply mimic/model/lstm_model.py]: I'm not sure about this. I do believe that we need a seed text (since LSTM predicts the next few words based on the current words I think?) but I'm not sure what we can pass in. We could look at just passing in some random text sequence from the corpus?

File: mimic/model/lstm_model.py (hunk 5 above)
Comment (frankfka) [reply mimic/model/lstm_model.py]: Just a plain string

File: mimic/model/lstm_model.py (hunk 1 above)
Comment (rwblickhan) [reply mimic/model/lstm_model.py]: I just merged #14, so you'll want to do this now. I would recommend rebasing on the latest version of master. In particular, in case you don't know how to do that (I should really have all of Launch Pad go through a rebase course just in case...):
* If you have any changes left on your local copy of this branch, commit, stash, or destroy them.
* Checkout out master and pull the latest version.
* Go back to this branch and do `git rebase master`.
* There shouldn't be any merge conflicts, but if there are, you can fix them, `git add` them, then `git rebase --continue` (you may have to do this repeatedly).
* You can now add more changes to the latest version of the branch!
* Finally, when you push, you'll have to do `git push -f` since the history of the branch has changed.

Comment (rwblickhan) [reply mimic/model/lstm_model.py]: In particular, you'll want to add Keras and TensorFlow under `[packages]` in `Pipfile`, then run `pipenv update --dev` to install all the dependencies and generate a new `Pipfile.lock`, which should be committed alongside `Pipfile` (there's slightly more detail in the [README](https://github.com/ubclaunchpad/mimic/blob/master/README.md)).

File: mimic/model/lstm_model.py (hunk 2 above)
Comment (rwblickhan) [reply mimic/model/lstm_model.py]: After you get `pipenv` set up, you'll want to run `pipenv run pycodestyle .` and `pipenv run pydocstyle .` - they'll probably complain about a lot of things, because they're quite picky! 😛 I do have Travis CI set up to prevent merging if they complain about anything, so you will have to do that before this can be merged.
